src/vsr/grid.zig

```diff
@@ -765,76 +892,57 @@ pub fn GridType(comptime Storage: type) type {
-            // resolves) to recovery queue. Future reads on the same address will see the "root"
-            // read in the recovery queue and enqueue to it.
-            grid.read_faulty_queue.push(read);
-            if (grid.on_read_fault) |on_read_fault| {
-                on_read_fault(grid, read);
-            } else {
-                @panic("Grid.on_read_fault not set");
-            }
+            grid.read_block_resolve(read, result);
         }
 
-        fn read_block_valid(block: BlockPtrConst, expect: struct {
+        fn read_block_validate(block: BlockPtrConst, expect: struct {
             address: u64,
             checksum: u128,
-            block_type: ?BlockType,
-        }) bool {
-            const address = expect.address;
-            const checksum = expect.checksum;
-            const block_type = expect.block_type;
-
+        }) ReadBlockResult {
             const header = mem.bytesAsValue(vsr.Header, block[0..@sizeOf(vsr.Header)]);
 
-            if (!header.valid_checksum()) {
-                log.err("invalid checksum at address {} (expected={})", .{ address, checksum });
-                return false;
-            }
+            if (!header.valid_checksum()) return .invalid_checksum;
+            if (header.command != .block) return .unexpected_command;
 
-            if (!header.valid_checksum_body(block[@sizeOf(vsr.Header)..header.size])) {
-                log.err("invalid checksum body at address {} (expected={})", .{ address, checksum });
-                return false;
-            }
+            assert(header.size >= @sizeOf(vsr.Header));
+            assert(header.size <= constants.block_size);
 
-            if (header.checksum != checksum) {
-                log.err(
-                    "expected address={} checksum={} block_type={?}, " ++
-                        "found address={} checksum={} block_type={}",
-                    .{
-                        address,
-                        checksum,
-                        block_type,
-                        header.op,
-                        header.checksum,
-                        @enumToInt(header.operation),
-                    },
-                );
-                return false;
-            }
+            const block_body = block[@sizeOf(vsr.Header)..header.size];
+            if (!header.valid_checksum_body(block_body)) return .invalid_checksum_body;
+            if (header.checksum != expect.checksum) return .unexpected_checksum;
 
-            assert(header.op == address);
-            if (block_type) |block_type_| {
-                assert(header.operation == block_type_.operation());
-            }
-            return true;
+            assert(header.op == expect.address);
+            return .{ .valid = block };
         }
 
-        fn read_block_resolve(grid: *Grid, read: *Grid.Read, block: BlockPtrConst) void {
+        fn read_block_resolve(grid: *Grid, read: *Grid.Read, result: ReadBlockResult) void {
```

Comment:
Previously `read_block_resolve` always received a `block`. Now it can receive a block _or_ an error (e.g. "invalid checksum").

If it receives a block, the behavior is unchanged.

If it receives a block, for each read in the resolve-chain:

- If the read originated from `read_block_from_replica` (`tag(callback) == .local`) report the error. (For the scrubber, this will trigger repair).
- If the read originated from `read_block_from_cluster` (`tag(callback) == .remote`) queue the read in `read_faulty_queue` and signal to the replica that we are waiting for a repair.